File: server/artwork_types.py
"""File-type allow-list shared with the TypeScript client and server.

The lists live in shared/artwork-types.json. Illustrator .ai and .eps are
vector artwork and follow the PDF path.
"""

import json
import os

_CATALOG_PATH = os.path.abspath(
    os.path.join(os.path.dirname(__file__), "..", "shared", "artwork-types.json")
)

with open(_CATALOG_PATH, "r", encoding="utf-8") as _catalog_file:
    _CATALOG = json.load(_catalog_file)

UPLOAD_EXTENSIONS = tuple(_CATALOG["uploadExtensions"])
PRINT_TOOL_EXTENSIONS = tuple(_CATALOG["printToolExtensions"])
VECTOR_EXTENSIONS = frozenset(_CATALOG["vectorExtensions"])
RASTER_EXTENSIONS = frozenset(_CATALOG["rasterExtensions"])
VECTOR_TYPES = frozenset(_CATALOG["vectorTypes"])
ILLUSTRATOR_TYPES = frozenset(_CATALOG["illustratorTypes"])


def _with_dot(value: str) -> str:
    raw = (value or "").strip().lower()
    if not raw:
        return ""
    return raw if raw.startswith(".") else f".{raw}"


def _bare(value: str) -> str:
    raw = (value or "").strip().lower().replace("\\", "/").split("/")[-1]
    if "." not in raw:
        return raw
    return raw.rsplit(".", 1)[-1]


def is_vector_extension(ext: str) -> bool:
    return _with_dot(ext) in VECTOR_EXTENSIONS


def is_raster_extension(ext: str) -> bool:
    return _with_dot(ext) in RASTER_EXTENSIONS


def is_vector_type(file_type: str) -> bool:
    return _bare(file_type) in VECTOR_TYPES


def is_illustrator_type(file_type: str) -> bool:
    return _bare(file_type) in ILLUSTRATOR_TYPES


def input_kind(ext: str) -> str:
    """How the press compiler should open this extension: pdf, image, or unsupported."""
    if is_vector_extension(ext):
        return "pdf"
    if is_raster_extension(ext):
        return "image"
    return "unsupported"

File: shared/artwork-types.ts
import catalog from "./artwork-types.json";

export const UPLOAD_EXTENSIONS = catalog.uploadExtensions;
export const PRINT_TOOL_EXTENSIONS = catalog.printToolExtensions;
export const VECTOR_EXTENSIONS = catalog.vectorExtensions;
export const RASTER_EXTENSIONS = catalog.rasterExtensions;
export const VECTOR_TYPES = catalog.vectorTypes;
export const ILLUSTRATOR_TYPES = catalog.illustratorTypes;

const DISPLAY_NAMES: Record<string, string> = catalog.displayNames;

function withDot(value: string): string {
  const raw = (value || "").trim().toLowerCase();
  if (!raw) return "";
  return raw.startsWith(".") ? raw : `.${raw}`;
}

function bareType(value: string): string {
  const base = (value || "").split(/[/\\]/).pop() || "";
  if (!base.includes(".")) return base.toLowerCase();
  return base.split(".").pop()?.toLowerCase() || "";
}

export function typeLabel(extensions: readonly string[]): string {
  const seen = new Set<string>();
  const parts: string[] = [];
  for (const ext of extensions) {
    if (ext === ".jpeg") continue;
    const name = DISPLAY_NAMES[ext] || ext.replace(/^\./, "").toUpperCase();
    if (seen.has(name)) continue;
    seen.add(name);
    parts.push(name);
  }
  return parts.join(", ");
}

export const ARTWORK_TYPE_LABEL = typeLabel(UPLOAD_EXTENSIONS);
export const PRINT_TOOL_TYPE_LABEL = typeLabel(PRINT_TOOL_EXTENSIONS);
export const PRINT_TOOL_ACCEPT = PRINT_TOOL_EXTENSIONS.join(",");

export function isUploadExtension(ext: string): boolean {
  return (UPLOAD_EXTENSIONS as readonly string[]).includes(withDot(ext));
}

export function isPrintToolExtension(ext: string): boolean {
  return (PRINT_TOOL_EXTENSIONS as readonly string[]).includes(withDot(ext));
}

export function isVectorExtension(ext: string): boolean {
  return (VECTOR_EXTENSIONS as readonly string[]).includes(withDot(ext));
}

export function isRasterExtension(ext: string): boolean {
  return (RASTER_EXTENSIONS as readonly string[]).includes(withDot(ext));
}

export function isPassThroughExtension(ext: string): boolean {
  return isVectorExtension(ext) || isRasterExtension(ext);
}

export function isVectorType(fileType: string): boolean {
  return (VECTOR_TYPES as readonly string[]).includes(bareType(fileType));
}

export function isIllustratorType(fileType: string): boolean {
  return (ILLUSTRATOR_TYPES as readonly string[]).includes(bareType(fileType));
}

export function isPrintToolType(fileType: string): boolean {
  return isPrintToolExtension(bareType(fileType));
}

export function isVectorArtwork(nameOrType: string): boolean {
  return isVectorType(nameOrType);
}

export function isIllustratorFile(nameOrType: string): boolean {
  return isIllustratorType(nameOrType);
}
